File: express/src/routes/Marketplace/trade_transactions.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { TradeTransactionService } from '../../services/Marketplace/trade_transaction_service.js';

const router = Router();
const service = new TradeTransactionService();

function validate(data: any): void {
  if ((data.status === 'Completed') && data.completedAt == null) throw new Error('completed_at is required');
  if (!((data.platformFee == null || (data.finalPrice != null && Number(data.platformFee) <= Number(data.finalPrice))))) throw new Error(`Platform fee cannot exceed the final price`);
  if (!((data.platformFee == null || Number(data.platformFee) >= 0))) throw new Error(`Platform fee must not be negative`);
  if (!((data.finalPrice == null || Number(data.finalPrice) > 0))) throw new Error(`Transaction final price must be greater than zero`);
  if ((data.status === 'COMPLETED') && !((data.completedAt === undefined || data.completedAt != null))) throw new Error(`Completed transaction must have a completed_at timestamp`);
}
function applyProjection(obj: any): any {
  if (!obj) return obj;
  const r = { ...obj };
  if ('completedAt' in r) { r.completedAt = r.completedAt; delete r.completedAt; }
  return r;
}


router.get('/', async (req, res) => {
  const items = await prisma.tradeTransaction.findMany();
  res.json(items.map(applyProjection));
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.tradeTransaction.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(applyProjection(entity));
});

router.post('/:id/complete', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.complete(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/refund', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.refund(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/dispute', async (req, res) => {
  const id = Number((req.params as any).id);
  const reason = req.body.reason;
  try {
    await service.open_dispute(id, reason);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.get('/:id/seller-net', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    const result = await service.seller_net(id);
    res.json({ result });
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});
export default router;

File: gin/internal/handler/marketplace/trade_transactions.go
package handler_marketplace

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/marketplace"
	"cards_project/internal/handler"
)

type TradeTransactionHandler struct { db *gorm.DB }

func NewTradeTransactionHandler(db *gorm.DB) *TradeTransactionHandler {
	return &TradeTransactionHandler{db: db}
}

func (h *TradeTransactionHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/trade_transactions")
	g.GET("", h.List)
	g.GET("/:id", h.Get)
	g.POST("/:id/api/transactions/{id}/complete", h.Complete)
	g.POST("/:id/api/transactions/{id}/refund", h.Refund)
	g.POST("/:id/api/transactions/{id}/dispute", h.OpenDispute)
	g.GET("/:id/api/transactions/{id}/seller-net", h.SellerNet)
}

func (h *TradeTransactionHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.TradeTransaction
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.TradeTransactionResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *TradeTransactionHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeTransaction
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeTransaction"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *TradeTransactionHandler) Complete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeTransaction
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeTransaction"); return }
		handler.DbError(c, err); return
	}
	err := row.Complete()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TradeTransactionHandler) Refund(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeTransaction
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeTransaction"); return }
		handler.DbError(c, err); return
	}
	err := row.Refund()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TradeTransactionHandler) OpenDispute(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeTransaction
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeTransaction"); return }
		handler.DbError(c, err); return
	}
	var body map[string]interface{}
	_ = c.ShouldBindJSON(&body)
	reason := func() string {
		v, ok := body["reason"]; if !ok { return "" }
		s, ok := v.(string); if !ok { return "" }
		return s
	}()
	err := row.OpenDispute(reason)
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TradeTransactionHandler) SellerNet(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeTransaction
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeTransaction"); return }
		handler.DbError(c, err); return
	}
	result, err := row.SellerNet()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

// ── Validation rules ─────────────────────────────────────────────
func validateTradeTransaction(req *model.TradeTransactionCreateRequest) []string {
	var errs []string
	if (req.Status == model.TradeTransactionStatusType_Completed) && req.CompletedAt == nil {
		errs = append(errs, "completed_at is required")
	}
	if !(float64(req.PlatformFee) <= float64(req.FinalPrice)) {
		errs = append(errs, "Platform fee cannot exceed the final price")
	}
	if !(float64(req.PlatformFee) >= 0) {
		errs = append(errs, "Platform fee must not be negative")
	}
	if !(float64(req.FinalPrice) > 0) {
		errs = append(errs, "Transaction final price must be greater than zero")
	}
	if !((!( req.Status == model.TradeTransactionStatusType_Completed ) || (req.CompletedAt != nil))) {
		errs = append(errs, "Completed transaction must have a completed_at timestamp")
	}
	return errs
}

func toCreateRequestTradeTransaction(m *model.TradeTransaction) model.TradeTransactionCreateRequest {
	return model.TradeTransactionCreateRequest{
		FinalPrice: m.FinalPrice,
		PlatformFee: m.PlatformFee,
		Status: m.Status,
		CompletedAt: m.CompletedAt,
		ListingID: m.ListingID,
		BuyerID: m.BuyerID,
		SellerID: m.SellerID,
	}
}
